File: lockerfit/optimizer.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta

from lockerfit.layout import LockerLayout
from lockerfit.models import ActiveSession, AssignmentRequest, CandidateOption
from lockerfit.predictor import DurationPredictor


@dataclass(frozen=True)
class AssignmentConfig:
    arrival_window_minutes: int = 10
    departure_window_minutes: int = 15
    passive_crowding_weight: float = 1.5
    overlap_weight: float = 85.0
    nearest_weight: float = 2.0
    average_near_weight: float = 0.8
    preferred_tier_bonus: float = 2.5
    avoided_tier_penalty: float = 7.5


@dataclass(frozen=True)
class TimeWindow:
    start: datetime
    end: datetime

    def overlap_minutes(self, other: "TimeWindow") -> float:
        start = max(self.start, other.start)
        end = min(self.end, other.end)
        if end <= start:
            return 0.0
        return (end - start).total_seconds() / 60.0

    @property
    def minutes(self) -> float:
        return max((self.end - self.start).total_seconds() / 60.0, 1.0)


class LockerAssigner:
    def __init__(
        self,
        layout: LockerLayout,
        predictor: DurationPredictor | None = None,
        config: AssignmentConfig | None = None,
    ) -> None:
        self.layout = layout
        self.predictor = predictor or DurationPredictor()
        self.config = config or AssignmentConfig()

    def recommend(
        self,
        request: AssignmentRequest,
        active_sessions: list[ActiveSession],
        *,
        top_k: int = 5,
        skip_locker_ids: set[int] | None = None,
        explain: bool = True,
    ) -> list[CandidateOption]:
        if top_k <= 0:
            raise ValueError("top_k must be positive")

        occupied = {session.locker_id for session in active_sessions}
        skipped = skip_locker_ids or set()
        candidate_ids = [
            locker_id
            for locker_id in self.layout.assignable_ids()
            if locker_id not in occupied and locker_id not in skipped
        ]

        expected_check_out = self.predictor.expected_checkout(
            request.arrived_at,
            request.visitor_hash,
        )
        candidate_windows = self._changing_windows(request.arrived_at, expected_check_out)
        session_profiles = [
            (
                session,
                self._temporal_overlap_weight(candidate_windows, self._session_windows(session)),
            )
            for session in active_sessions
        ]

        scored = [
            self._score_candidate(
                locker_id,
                request,
                expected_check_out,
                candidate_windows,
                session_profiles,
                include_reasons=False,
            )
            for locker_id in candidate_ids
        ]
        scored.sort(key=lambda option: (-option.score, option.locker_id))
        top = scored[:top_k]
        if not explain:
            return top

        return [
            self._score_candidate(
                option.locker_id,
                request,
                expected_check_out,
                candidate_windows,
                session_profiles,
                include_reasons=True,
            )
            for option in top
        ]

    def assign(
        self,
        request: AssignmentRequest,
        active_sessions: list[ActiveSession],
        *,
        skip_locker_ids: set[int] | None = None,
        explain: bool = True,
    ) -> CandidateOption:
        options = self.recommend(
            request,
            active_sessions,
            top_k=1,
            skip_locker_ids=skip_locker_ids,
            explain=explain,
        )
        if not options:
            raise RuntimeError("no available lockers")
        return options[0]

    def _score_candidate(
        self,
        locker_id: int,
        request: AssignmentRequest,
        expected_check_out: datetime,
        candidate_windows: tuple[TimeWindow, TimeWindow],
        session_profiles: list[tuple[ActiveSession, float]],
        *,
        include_reasons: bool,
    ) -> CandidateOption:
        locker = self.layout.get(locker_id)
        distances = [
            (session, temporal_weight, self.layout.distance(locker_id, session.locker_id))
            for session, temporal_weight in session_profiles
        ]
        distances.sort(key=lambda item: item[2])

        nearest_distance = distances[0][2] if distances else None
        average_near = self._average_nearest(distances, limit=5)
        passive_crowding = sum(1.0 / (distance + 1.0) for _, _, distance in distances)
        overlap_risk = sum(
            self._overlap_risk(temporal_weight, distance)
            for _, temporal_weight, distance in distances
        )

        score = 0.0
        if nearest_distance is not None:
            score += nearest_distance * self.config.nearest_weight
        if average_near is not None:
            score += average_near * self.config.average_near_weight

        score -= passive_crowding * self.config.passive_crowding_weight
        score -= overlap_risk * self.config.overlap_weight

        if request.preferred_tier and locker.tier == request.preferred_tier:
            score += self.config.preferred_tier_bonus
        if locker.tier in request.avoided_tiers:
            score -= self.config.avoided_tier_penalty

        return CandidateOption(
            locker_id=locker_id,
            score=round(score, 4),
            nearest_active_distance=None if nearest_distance is None else round(nearest_distance, 3),
            overlap_risk=round(overlap_risk, 4),
            expected_check_out=expected_check_out,
            reasons=(
                self._reasons(locker_id, distances, overlap_risk, request)
                if include_reasons
                else ()
            ),
        )

    def _overlap_risk(
        self,
        temporal_weight: float,
        distance: float,
    ) -> float:
        return temporal_weight * (1.0 / (distance + 1.0))

    def _temporal_overlap_weight(
        self,
        candidate_windows: tuple[TimeWindow, TimeWindow],
        session_windows: tuple[TimeWindow, TimeWindow],
    ) -> float:
        overlap = 0.0
        for candidate_window in candidate_windows:
            for session_window in session_windows:
                overlap += candidate_window.overlap_minutes(session_window) / candidate_window.minutes
        return overlap

    def _changing_windows(
        self,
        check_in: datetime,
        expected_check_out: datetime,
    ) -> tuple[TimeWindow, TimeWindow]:
        arrival_end = check_in + timedelta(minutes=self.config.arrival_window_minutes)
        departure_start = expected_check_out - timedelta(minutes=self.config.departure_window_minutes)
        if departure_start < arrival_end:
            departure_start = arrival_end
        return (
            TimeWindow(check_in, arrival_end),
            TimeWindow(departure_start, expected_check_out),
        )

    def _session_windows(self, session: ActiveSession) -> tuple[TimeWindow, TimeWindow]:
        session_checkout = session.expected_check_out or self.predictor.expected_checkout(
            session.check_in,
            session.visitor_hash,
        )
        return self._changing_windows(session.check_in, session_checkout)

    @staticmethod
    def _average_nearest(
        distances: list[tuple[ActiveSession, float, float]],
        *,
        limit: int,
    ) -> float | None:
        if not distances:
            return None
        subset = [distance for _, _, distance in distances[:limit]]
        return sum(subset) / len(subset)

    def _reasons(
        self,
        locker_id: int,
        distances: list[tuple[ActiveSession, float, float]],
        overlap_risk: float,
        request: AssignmentRequest,
    ) -> tuple[str, ...]:
        locker = self.layout.get(locker_id)
        reasons: list[str] = [f"{locker.zone}, {locker.tier} tier"]
        if distances:
            nearest_session, _, nearest_distance = distances[0]
            reasons.append(
                f"nearest occupied locker is {nearest_session.locker_id} "
                f"at distance {nearest_distance:.1f}"
            )
        else:
            reasons.append("no occupied lockers in the current layout")
        if overlap_risk < 0.05:
            reasons.append("low predicted changing-window overlap")
        else:
            reasons.append(f"predicted overlap risk {overlap_risk:.2f}")
        if request.preferred_tier and locker.tier == request.preferred_tier:
            reasons.append(f"matches preferred tier {request.preferred_tier}")
        return tuple(reasons)

File: typescript/src/optimizer.ts
import { LockerLayout } from "./layout.js";
import type { ActiveSession, AssignmentRequest, CandidateOption } from "./types.js";

interface TimeWindow {
  start: Date;
  end: Date;
}

export interface AssignmentConfig {
  arrivalWindowMinutes: number;
  departureWindowMinutes: number;
  passiveCrowdingWeight: number;
  overlapWeight: number;
  nearestWeight: number;
  averageNearWeight: number;
  preferredTierBonus: number;
  avoidedTierPenalty: number;
}

const defaultConfig: AssignmentConfig = {
  arrivalWindowMinutes: 10,
  departureWindowMinutes: 15,
  passiveCrowdingWeight: 1.5,
  overlapWeight: 85,
  nearestWeight: 2,
  averageNearWeight: 0.8,
  preferredTierBonus: 2.5,
  avoidedTierPenalty: 7.5,
};

export class LockerAssigner {
  constructor(
    private readonly layout: LockerLayout,
    private readonly config: AssignmentConfig = defaultConfig,
  ) {}

  recommend(
    request: AssignmentRequest,
    activeSessions: ActiveSession[],
    topK = 5,
    skipLockerIds = new Set<number>(),
  ): CandidateOption[] {
    const occupied = new Set(activeSessions.map((session) => session.lockerId));
    const expectedCheckOut = addMinutes(
      request.arrivedAt,
      request.expectedDurationMinutes ?? 90,
    );
    const candidateWindows = this.changingWindows(request.arrivedAt, expectedCheckOut);
    const sessionProfiles = activeSessions.map((session) => ({
      session,
      temporalWeight: this.temporalOverlapWeight(
        candidateWindows,
        this.changingWindows(
          session.checkIn,
          session.expectedCheckOut ?? addMinutes(session.checkIn, 90),
        ),
      ),
    }));

    return this.layout
      .assignableIds()
      .filter((lockerId) => !occupied.has(lockerId) && !skipLockerIds.has(lockerId))
      .map((lockerId) =>
        this.scoreCandidate(lockerId, request, expectedCheckOut, sessionProfiles),
      )
      .sort((a, b) => b.score - a.score || a.lockerId - b.lockerId)
      .slice(0, topK);
  }

  private scoreCandidate(
    lockerId: number,
    request: AssignmentRequest,
    expectedCheckOut: Date,
    sessionProfiles: Array<{ session: ActiveSession; temporalWeight: number }>,
  ): CandidateOption {
    const locker = this.layout.get(lockerId);
    const distances = sessionProfiles
      .map((profile) => ({
        session: profile.session,
        temporalWeight: profile.temporalWeight,
        distance: this.layout.distance(lockerId, profile.session.lockerId),
      }))
      .sort((a, b) => a.distance - b.distance);

    const nearest = distances.length > 0 ? distances[0].distance : null;
    const near = distances.slice(0, 5).map((item) => item.distance);
    const averageNear = near.length > 0 ? near.reduce((a, b) => a + b, 0) / near.length : null;
    const passiveCrowding = distances.reduce(
      (total, item) => total + 1 / (item.distance + 1),
      0,
    );
    const overlapRisk = distances.reduce(
      (total, item) => total + this.overlapRisk(item.temporalWeight, item.distance),
      0,
    );

    let score = 0;
    if (nearest !== null) score += nearest * this.config.nearestWeight;
    if (averageNear !== null) score += averageNear * this.config.averageNearWeight;
    score -= passiveCrowding * this.config.passiveCrowdingWeight;
    score -= overlapRisk * this.config.overlapWeight;
    if (request.preferredTier && locker.tier === request.preferredTier) {
      score += this.config.preferredTierBonus;
    }
    if (request.avoidedTiers?.includes(locker.tier)) {
      score -= this.config.avoidedTierPenalty;
    }

    const reasons = [
      `${locker.zone}, ${locker.tier} tier`,
      nearest === null
        ? "no occupied lockers in the current layout"
        : `nearest occupied locker is ${distances[0].session.lockerId} at distance ${nearest.toFixed(1)}`,
      overlapRisk < 0.05
        ? "low predicted changing-window overlap"
        : `predicted overlap risk ${overlapRisk.toFixed(2)}`,
    ];

    return {
      lockerId,
      score: round(score, 4),
      nearestActiveDistance: nearest === null ? null : round(nearest, 3),
      overlapRisk: round(overlapRisk, 4),
      expectedCheckOut,
      reasons,
    };
  }

  private overlapRisk(
    temporalWeight: number,
    distance: number,
  ): number {
    return temporalWeight * (1 / (distance + 1));
  }

  private temporalOverlapWeight(
    candidateWindows: [TimeWindow, TimeWindow],
    sessionWindows: [TimeWindow, TimeWindow],
  ): number {
    let overlap = 0;

    for (const candidateWindow of candidateWindows) {
      for (const sessionWindow of sessionWindows) {
        overlap += overlapMinutes(candidateWindow, sessionWindow) / minutes(candidateWindow);
      }
    }

    return overlap;
  }

  private changingWindows(checkIn: Date, expectedCheckOut: Date): [TimeWindow, TimeWindow] {
    const arrivalEnd = addMinutes(checkIn, this.config.arrivalWindowMinutes);
    let departureStart = addMinutes(expectedCheckOut, -this.config.departureWindowMinutes);
    if (departureStart < arrivalEnd) {
      departureStart = arrivalEnd;
    }
    return [
      { start: checkIn, end: arrivalEnd },
      { start: departureStart, end: expectedCheckOut },
    ];
  }
}

function addMinutes(date: Date, minutesToAdd: number): Date {
  return new Date(date.getTime() + minutesToAdd * 60_000);
}

function overlapMinutes(a: TimeWindow, b: TimeWindow): number {
  const start = Math.max(a.start.getTime(), b.start.getTime());
  const end = Math.min(a.end.getTime(), b.end.getTime());
  return end <= start ? 0 : (end - start) / 60_000;
}

function minutes(window: TimeWindow): number {
  return Math.max((window.end.getTime() - window.start.getTime()) / 60_000, 1);
}

function round(value: number, precision: number): number {
  const multiplier = 10 ** precision;
  return Math.round(value * multiplier) / multiplier;
}
